src/Raven.Server/ServerWide/ServerStore.cs

```diff
@@ -99,10 +101,73 @@ public long LastRaftCommitEtag
 
         private Timer _timer;
         private RachisConsensus<ClusterStateMachine> _engine;
-        private bool _disposed;
-
+        private bool _disposed;      
         public RachisConsensus<ClusterStateMachine> Engine => _engine;
 
+        private ClusterMaintenanceMaster _clusterMaintenanceMaster;
+        public Dictionary<string, ClusterNodeStatusReport> ClusterStats()
+        {
+            if (_engine.LeaderTag != NodeTag)
+                return null; // not a leader
+            return _clusterMaintenanceMaster?.GetStats();
+        }
+        public async Task ClusterMaintanceSetupTask()
+        {
+            while (true)
+            {
+                try
+                {
+                    if (_engine.LeaderTag != NodeTag)
+                    {
+                        await _engine.WaitForState(RachisConsensus.State.Leader);
+                        continue;
+                    }
+                    using (_clusterMaintenanceMaster = new ClusterMaintenanceMaster(_engine.Tag, _engine.CurrentTerm))
+                    {
+                        var oldNodes = new Dictionary<string, string>();
+                        while (_engine.LeaderTag == NodeTag)
+                        {
+                            var topologyChangedTask = _engine.GetTopologyChanged();
+                            ClusterTopology clusterTopology;
+                            using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+                            using (context.OpenReadTransaction())
+                            {
+                                clusterTopology = _engine.GetTopology(context);
+                            }
+                            var newNodes = clusterTopology.AllNodes;
+                            var nodesChanges = ClusterTopology.DictionaryDiff(oldNodes, newNodes);
+                            oldNodes = newNodes;
+                            foreach (var node in nodesChanges.removedValues)
+                            {
+                                _clusterMaintenanceMaster.RemoveFromCluster(node.Key);
+                            }
+                            var taskList = new List<Task>();
+                            foreach (var node in nodesChanges.addedValues)
+                            {
+                                var task = _clusterMaintenanceMaster.AddToCluster(node.Key, clusterTopology.GetUrlFromTag(node.Key));
+                                taskList.Add(task);
+                            }
+
+                            var leaderChanged = _engine.WaitForLeaveState(RachisConsensus.State.Leader);
+                            var allNodesAdded = Task.WhenAll(taskList);
+                            if (await Task.WhenAny(allNodesAdded, topologyChangedTask, leaderChanged) == leaderChanged)
```

Comment:
I don't follow what you are doing here, why are you waiting for the `AddToCluster` result here?